cql3/statements/alter_keyspace_statement.cc

```diff
@@ -130,6 +130,64 @@ bool cql3::statements::alter_keyspace_statement::changes_tablets(query_processor
     return ks.get_replication_strategy().uses_tablets() && !_attrs->get_replication_options().empty();
 }
 
+namespace {
+// These functions are used to flatten all the options in the keyspace definition into a single-level map<string, string>.
+// (Currently options are stored in a nested structure that looks more like a map<string, map<string, string>>).
+// Flattening is simply joining the keys of maps from both levels with a colon ':' character,
+// or in other words: prefixing the keys in the output map with the option type, e.g. 'replication', 'storage', etc.,
+// so that the output map contains entries like: "replication:dc1" -> "3".
+// This is done to avoid key conflicts and to be able to de-flatten the map back into the original structure.
+
+void add_prefixed_key(const sstring& prefix, const std::map<sstring, sstring>& in, std::map<sstring, sstring>& out) {
+    for (const auto& [in_key, in_value]: in) {
+        out[prefix + ":" + in_key] = in_value;
+    }
+};
+
+std::map<sstring, sstring> get_current_options_flattened(const shared_ptr<cql3::statements::ks_prop_defs>& ks,
+                                                         bool include_tablet_options,
+                                                         const gms::feature_service& feat) {
+    std::map<sstring, sstring> all_options;
+
+    add_prefixed_key(ks->KW_REPLICATION, ks->get_replication_options(), all_options);
+    add_prefixed_key(ks->KW_STORAGE, ks->get_storage_options().to_map(), all_options);
+    add_prefixed_key(ks->KW_TABLETS, ks->get_map(ks->KW_TABLETS).value_or(std::map<sstring, sstring>{}), all_options);
+    // if no tablet options are specified in ATLER KS statement,
+    // we want to preserve the old ones and hence cannot overwrite them with defaults
+    if (include_tablet_options) {
+        auto initial_tablets = ks->get_initial_tablets("NetworkTopologyStrategy", feat.tablets);
+        add_prefixed_key(ks->KW_TABLETS,
+                         {{"enabled", initial_tablets.enabled ? "true" : "false"},
+                         {"initial", std::to_string(initial_tablets.specified_count.value_or(0))}},
```

Comment:
If initial_tablets.enabled == false why do we need to set `initial` to 0?

Also, if `ks->get_map(ks->KW_TABLETS)` is disengaged, why do we bother to call `add_prefixed_key` with an empty map?  Wouldn't it be simpler to check the map existence (maybe even inside this block) and call `add_prefixed_key` only if `ks->get_map(ks->KW_TABLETS)` is engaged?

Reply:
I see below that `bool include_tablet_options = _attrs->get_map(_attrs->KW_TABLETS).has_value()`.
Since we cannot ALTER a keyspace from non-tablets to tablets, the combination of `include_tablet_options` and `!ks->get_map(ks->KW_TABLETS).has_value()` should probably trigger an error.

Reply:
> If initial_tablets.enabled == false why do we need to set initial to 0?

If that's the case, then we don't need to set `initial`, but it doesn't hurt to have this value set, as it's not going to be used anyways. We do set it nevertheless to have `initial` set if `initial_tablets.enabled == true`. So this is really just to make the code shorter.

I'll reply to other comments in a while.

Reply:
Adding tablets options 2 times:
```
    add_prefixed_key(ks->KW_TABLETS, ks->get_map(ks->KW_TABLETS).value_or(std::map<sstring, sstring>{}), all_options);
    // if no tablet options are specified in ATLER KS statement,
    // we want to preserve the old ones and hence cannot overwrite them with defaults
    if (include_tablet_options) {
        auto initial_tablets = ks->get_initial_tablets("NetworkTopologyStrategy", feat.tablets);
        add_prefixed_key(ks->KW_TABLETS,
                         {{"enabled", initial_tablets.enabled ? "true" : "false"},
                         {"initial", std::to_string(initial_tablets.specified_count.value_or(0))}},
                         all_options);
    }
```
was my mistake. The code under the `if` was supposed to replace the 1st line, but somehow I left both. If I remove the 1st line, all test.py testcases pass, which proves the 1st line was redundant.

Reply:
I ended up removing this redundant check.